stbt/android.py

```diff
@@ -0,0 +1,406 @@
+        `Stb-tester CAMERA`_, or ``ADB_NATIVE`` elsewhere.
+    """
+
+    def __init__(self, adb_server=None, adb_device=None, adb_binary="adb",
+                 tcpip=False, coordinate_system=None):
+        self.adb_server = adb_server
+        self._adb_device = adb_device
+        self.adb_binary = adb_binary
+        self.tcpip = tcpip
+        self.coordinate_system = coordinate_system
+
+    @property
+    def adb_device(self):
+        if self.tcpip and self._adb_device and ":" not in self._adb_device:
+            return self._adb_device + ":5555"
+        else:
+            return self._adb_device
+
+    def adb(self, command, timeout_secs=5 * 60, capture_output=False, **kwargs):
+        """Run any adb command.
+
+        ``command`` and ``kwargs`` are the same as `subprocess.check_output`,
+        except that ``shell``, ``stdout`` and ``stderr`` are not allowed.
+
+        Raises `AdbError` if the command fails.
+        """
+        try:
+            if self.tcpip:
+                self._connect(timeout_secs)
+            output = self._adb(command, timeout_secs, **kwargs)
+        except subprocess.CalledProcessError as e:
+            raise AdbError(e.returncode, e.cmd, e.output, self), \
+                None, sys.exc_info()[2]
+        if capture_output:
+            return output
+        else:
+            sys.stderr.write(output)
+            return None
+
+    def devices(self):
+        try:
+            return self._adb(["devices", "-l"], timeout_secs=5)
+        except subprocess.CalledProcessError as e:
+            return e.output
+
+    def screenshot(self):
+        """Take a screenshot over adb, in the same orientation that the device
+        is in."""
+
+        for attempt in range(1, 11):
+            data = (self.adb(["shell", "screencap", "-p"],
+                             timeout_secs=60, capture_output=True)
+                    .replace("\r\n", "\n"))
+            img = cv2.imdecode(
+                numpy.asarray(bytearray(data), dtype=numpy.uint8),
+                cv2.CV_LOAD_IMAGE_COLOR)
+            if img is None:
+                sys.stderr.write(
+                    "AdbDevice.screenshot: error: Failed to get screenshot "
+                    "via adb (attempt %d/10)\n" % attempt)
+                sys.stderr.write("Length of data: %d\n" % len(data))
+                sys.stderr.write("Data: %r\n" % data)
+            else:
+                break
+        else:
+            raise RuntimeError(
+                "Failed to capture screenshot from android device")
+
+        w, h = img.shape[1], img.shape[0]
+        stbt.debug("Screenshot is %d px wide x %d px high" % (w, h))
+
+        if self.coordinate_system == CoordinateSystem.ADB_720P:
+            # Resize to 720p preserving orientation
+            if w > h:
+                img = cv2.resize(img, (1280, 720))
+            else:
+                img = cv2.resize(img, (720, 1280))
+
+        return img
```

Comment:
> Would there be any value in returning a stbt.Frame here

Good idea. Done.

I did a quick test by starting the Android stopwatch when my computer ticked over to 0 seconds of a new minute, and printed the time before & after calling `AdbDevice.screenshot` as well as saving the frame. Seems to be within 0.1s variance. For now I've documented it as "probably to <0.5s or so". To quantify it we could capture frames from HDMI and from ADB at the same time, and do OCR on the time on the screen. But that's outside the scope of this pull request.